byterun/domain.c

```diff
@@ -324,16 +366,43 @@ CAMLprim value caml_domain_spawn(value callback)
   if (p.newdom) {
     /* successfully created a domain.
        p.callback is now owned by that domain */
-    pthread_detach(th);
   } else {
     /* failed */
-    void* r;
-    pthread_join(th, &r);
+    caml_ml_domain_join(th_val);
     caml_delete_root(p.callback);
     caml_failwith("failed to allocate domain");
   }
 
-  CAMLreturn (Val_unit);
+  CAMLreturn (th_val);
+}
+
+CAMLprim value caml_ml_domain_join(value domain)
+{
+    CAMLparam1 (domain);
+    struct domain_thread **dp = Domainthreadptr_val(domain);
+    struct domain_thread *d;
+    int err = 0;
+    pthread_t th;
+
+    Assert(dp);
+    d = *dp;
+    Assert(d);
+    th = d->th;
```

Comment:
Here, `d->th` is accessed without holding the mutex `d->m`. It's not wrong, because `d->th` is readonly except in the finaliser, with which no code can race. However, the locking setup isn't obvious.

You should do one of:
  - Move this line after `caml_plat_lock` below so that `d->m` protects all accesses to `d`, or
  - Add a comment in the definition of `struct domain_thread` explaining that the mutex protects only `joinable` but not `th`, and that `th` is immutable once initialised.